### PR TITLE
Make branch deletion URL more like GitHub's, fixes #1397

### LINKED ISSUE #1397: Cannot delete branch with slash in name after PR
- Gitea version (or commit ref):1.1.0
- Git version:
- Operating system: Ubuntu
- Database (use `[x]`):
  - [ ] PostgreSQL
  - [x] MySQL
  - [ ] MSSQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [x] Yes (provide example URL) https://try.gitea.io/username/branch-slash/pulls/1
  - [ ] No
  - [ ] Not relevant
- Log gist:
https://gist.github.com/Aarup/75908798b147e91ffb1133357acb44a1

## Description
If a branch has has a slash in the name like
  - bug/bugsId
  - feature/featureName
It is impossible to delete this branch using the web UI after a PR has been merged into the repository.
When clicking the "Delete branch" button a generic popup comes, but after entering username and password nothing happens - a 404 is received from the post

## Screenshots
image
image

### PR DESCRIPTION
**Pull request by lafriks**
Also fixes broken branch deletion as MergeCommitID is not in HeadBranch so it allways was giving error message that branch has new commits and it can not be deleted.

Fixes #1397

### CODE REVIEW

**Comment by lunny** [comment]
Maybe add an integration test since merge test is ready.

File: routers/repo/pull.go
Comment (ethantkoenig) [inline routers/repo/pull.go]: Should it be `ctx.Repo.Repository.DefaultBranch` instead of `"master"`?

File: routers/repo/pull.go
Comment (ethantkoenig) [inline routers/repo/pull.go]: 403 seems better than 404 here (although I'm no expert, could be wrong).

**Review by ethantkoenig — CHANGES_REQUESTED** [review]
Will this URL change prevent us from being able to support deleting arbitrary branches (even those not associated with any PR) in the future?

File: routers/repo/pull.go
Comment (lunny) [reply routers/repo/pull.go]: No, 404 is better.

**Comment by lafriks** [comment]
Changed to checking for default repository branch name instead of "master"

**Comment by lafriks** [comment]
This URL meant to be used only for deleting PR branch and does very specific checks and functionality. For branch deletion from other places it would be better to create new URL with new handler

**Comment by lunny** [comment]
When I try to delete the merged pull branch, it shows `lunny/test2 cannot be deleted because it has new commits after merging.`

**Comment by lafriks** [comment]
@lunny did you do it for fork? Does it works on try gitea? If both answers are yes, than you are hitting this bug #2007

File: routers/repo/pull.go
Comment (bkcsoft) [reply routers/repo/pull.go]: Yes

File: routers/repo/pull.go
Comment (bkcsoft) [reply routers/repo/pull.go]: If you can see the PR you should get a 403, otherwise 404

Comment (lafriks) [reply routers/repo/pull.go]: I think it will check for PR view rights in routes.go already

Comment (bkcsoft) [reply routers/repo/pull.go]: Then 403 is correct

**Comment by lafriks** [comment]
Added integration test does fail without #2007 applied, so it covers both my PR :)

**Comment by lunny** [comment]
The close comment and deletion comment is not shown but I think that should be another issue. So I give this LGTM.

**Comment by lunny** [comment]
Another thinking: if pull branch is a protected branch?

File: integrations/pull_merge_test.go
Comment (ethantkoenig) [inline integrations/pull_merge_test.go]: Unchecked error, I'd recommend using the `DecodeJSON(..)` helper function here:

```
DecodeJSON(t, resp, respJSON)
```

**Review by ethantkoenig — CHANGES_REQUESTED** [review]

**Comment by lafriks** [comment]
@ethantkoenig fixed

**Comment by lafriks** [comment]
@lunny added checks to not allow deleting protected branches

**Comment by lafriks** [comment]
Missing comments are bug introduced in 255adc40ae1e0a03547326f143f685be41fa591f and needs to be fixed in other PR

**Comment by lafriks** [comment]
@ethantkoenig need your approval

**Comment by ethantkoenig** [comment]
@lafriks Still need to address https://github.com/go-gitea/gitea/pull/1994#discussion_r122571873

**Comment by lafriks** [comment]
@ethantkoenig ok, done

**Review by ethantkoenig — APPROVED** [review]

**Comment by ethantkoenig** [comment]
LGTM

**Comment by lafriks** [comment]
@lunny @ethantkoenig I added additional check so that code would not panic if forked repository has been deleted.
There are some other places where it does panic but for them I will create new PR

**Pull request merged**
